### ISSUE TITLE
Trying it out on Arch Linux

### ISSUE THREAD

**Issue by cadr10** [T-vK/MobilePassThrough#8]
I've successfully made qemu work in arch and setup kvm. Came across your program and setup everything as describe in the readme. When I run ./start-vm.sh, both my monitors go black (this happens after unbinding the pci). Any idea why it's happening? I use optimus-manager instead of bumblebee and I tried setting mode to intel, hybrid and nvidia. Same result.

Device: Aspire VX5-591G (muxed, according to compability-check). 

I'll try to fix it myself in the coming days but posting this if there's clarity somewhere.

**Comment by T-vK** [comment]
At the moment only Fedora is supported. If you want to make it work on Arch, you have create utils/Arch/{version} and translate the files from https://github.com/T-vK/MobilePassThrough/tree/master/utils/Fedora/30 to work on Arch.

**Comment by cadr10** [comment]
What are the dependencies to generate the helper iso? I ran the script to generate it and it just downloads virtio-drivers and couple of exes. 
I got all of the dependencies with official repositories and aur. When I reinstall my os, I'll make a list of what I installed to get it working.
I know question is different topic but I am not opening a new issue for it, l.

**Comment by T-vK** [comment]
For the helper iso you need a package providing genisoimage. Not sure which version.

**Comment by Rabcor** [comment]
> For the helper iso you need a package providing genisoimage. Not sure which version.
> I haven't actually tested the iso generator in quite some time. It might be broken.
> 
> I think the last line should probably be replaced with
> 
> ```
> genisoimage -J -joliet-long -r -allow-lowercase -allow-multidot -o "${HELPER_ISO}" "${PROJECT_DIR}/helper-iso-files"
> ```

Yeah I'm trying to set this up on manjaro over here, the original files ended up giving me these errors

```
> Generating new iso...
Warning: creating filesystem that does not conform to ISO-9660.
Setting input-charset to 'UTF-8' from locale.
genisoimage: No such file or directory. Invalid node - '/home/rabcor/Windoze/vm-files/mobile-passthrough-helper.iso'.
```
When running the genisoimage command, but replacing it with the version you mentioned in this comment solved it.

NOw I'm finally at the start vm stage, crossing my fingers. I didn't actually fix the dependency download scripts though, I just installed them all manually, so unfortunately I have no such files to share, but any arch user should easily be able to figure out what to download from the files.

Also depends on cdrtools for aforementioned genisoimage command.

**Comment by Rabcor** [comment]
Ok so I had to create a user (useradd -m qemu) then run those looking-glass-setup script commands to give it permission to use looking glass to get rid of the missing user error; additionally I had to edit where start-vm.sh tries to use optirun and make it use primusrun instead, because optirun was not working correctly.

And yea after fixing everything I ran start-vm.sh > output.txt to try and see what happens when I get the black screen, and this is it:

```
> Loading config from /home/rabcor/Windoze/user.conf
> Using a virtual OS drive...
> Bumblebee works fine on this system. Using optirun when necessary...
> Retrieving and parsing DGPU IDs...
> Loading vfio-pci kernel module...
> Using Looking Glass...
> Calculating required buffer size for 1920x1080 for Looking Glass...
> Looking Glass buffer size set to: 32M
> Starting IVSHMEM server...
> Adjusting permissons for the IVSHMEM server socket...
> Not using DGPU vBIOS override...
> Using SMB share...
> Using dGPU passthrough...
> Unbinding dGPU from nvidia driver...
0000:01:00.0 /sys/bus/pci/devices/0000:01:00.0/driver/unbind
> Binding dGPU to VFIO driver...
> Using mediated iGPU passthrough...
> Create vGPU for mediated iGPU passthrough...
> Using dma-buf...
> Using spice on port 5900...
> Using QXL...
> Not using USB passthrough...
> Using virtual input method 'usb-tablet' for keyboard/mouse input...
> Starting the Virtual Machine...
QEMU 5.1.0 monitor - type 'help' for more information
(qemu) DNSMASQ terminated
ip_forward disabled
> Binding dGPU back to nvidia driver...
> Remove Intel vGPU...
```

not entirely sure what to make of it.

**Comment by T-vK** [comment]
Sorry for all the bugs. I haven't done any work on this project in a long time. It's great to see though that you figured out most of the issues. 
I remember having a lot of issues with looking glass as well the last time I was working on this. RDP did work pretty reliably for me though. The only issue with it that I can remember was that I had to set up a password on the Windows user account, otherwise it wouldn't let me connect.

**Comment by Rabcor** [comment]
@T-vK well I knew what to expect for a project that hasn't been updated for almost 2 years ;)

I'm surprised it worked as well as it did in fact. I probably never could have figured out the initial setup for qemu without your script automating and guiding me through most of the process (I've tried before, so I know I wouldn't even get halfway through it in fact)

I'm still stumped by the RDP issues but I'm too burnt out from working all day yesterday to get to the point I'm currently at to go troubleshoot that too (besides I don't even know where to start, I tried it, got some errors, googled the errors, and got essentially nothing, I also tried xfreerdp only to get the exact same results, failure to connect)

It sucks to be so close, and then get caught on some stupid technicality like this that was never even supposed to be a possible issue :frowning_face: it looks like everything should work, but it just doesn't, I'm so close I can taste it, I even managed to get rid of that legendary code 43 that trips up most people (I was actually expecting to get stuck there, not... not here...)

I also have problems with ACPI on the host, namely that bbswitch can't seem to properly turn off the dgpu, I also tried this acpi_call method that's supposed to do the same but I always get a permission denied error there; unless I solve this issue I won't be able to turn the VM on at any point after using the dGPU on the host, if I solve it however, it might open a lot of doors.

As it stands it looks like these are the only two issues left, but I'm just too burnt out to deal with them today, it's a cold winter night, so I'm gonna make hot chocolate and watch the next episode of mandalorian instead :+1: 

Hopefully tomorrow's me will be able to figure something out.

**Comment by Rabcor** [comment]
I've done some fixes to the start-vm.sh, fixed a few important typos/broken commands. (Like unbinding the dgpu from the nvidia driver, and the subsystem id collection)

This allows me to successfully run the VM again after using the dGPU on the host machine, however unbinding the dGPU from vfio-pci is no longer as easy as optirun doing it automagically, and the commands in the script to do just that (the ones which formerly broke my PC) are not doing it right either.

Additionally since bumblebee is kinda obsolete and I'm having some problems with it I was wondering if I could do this with Prime instead, I can actually switch between prime and bumblebee as simply as running 1 command and restarting the machine, but when I tried with prime the script hangs on unbinding the dGPU from the driver.

So I have 3 goals now.

1: Get RDP working :crying_cat_face: 
2: Reconnect dGPU to host nvidia driver after shutting down VM successfully
3: Get everything working with Prime instead of Bumblebee (it seems there's only something minor missing for this; I know with the older method of using prime to run X from nvidia or intel instead of allowing both to co-exist, it was possible to swap between the two without a restart, although it did require logging out and back in, I might be able to find clues there)

...

Reconnecting the dGPU was quite a simple matter, all I had to do was do it manually, I added this line to the script:
`sudo bash -c "echo '${DGPU_VENDOR_ID} ${DGPU_DEVICE_ID}' > '/sys/bus/pci/drivers/nvidia/new_id'"`

And that does it, I can now freely unbind/rebind the dGPU to the nvidia/vfio-pci drivers, and start/shut down the VM as many times as I want without any issues besides the aforementioned RDP ones.

The fact that I'm this close to success but am still failing to get RDP working and thus cannot test if the dGPU is actually working inside the VM or not is maddening. It appears that everything except RDP is now working near-flawlessly.

**Comment by T-vK** [comment]
I haven't really looked into prime, but from what I recall it didn't allow for switching GPUs without logging out. Having to log out makes for a really bad user experience imo. I was also hoping to support the nouveau driver at some point.
Maybe it would make sense to add a config option to let the user decide if he wants to use bumblebee, nvidia-prime or nothing at all.

It's really great to see that you found and fixed so many bugs. Maybe you can make a pull request so that others can benefit in the future.

Can you go a bit more into detail on your RDP issue? 
Have you checked if you can ping the VM at all from the host? 
Did you check if the RDP port is actually exposed properly (e.g. using nmap)? 
If you have a USB network card maybe you can try passing that through directly. I think that's how I ended up using my VM most of the time. 

I'm not sure about audio issues. Last time I was working on this my focus was on getting the video part working more stable. I haven't spent any time focusing on audio or performance optimizations. 
Pulseaudio on my Fedora machines has never really been stable to begin with so I probably wouldn'T have noticed any additional issues. In regards to audio within the VM I always used USB passthrough to pass a USB sound card through to the VM.

**Comment by Rabcor** [comment]
I finally got freerdp to work, apparently the problem was with the Guest machine's firewall settings, turning off the firewall worked as a hotfix, gonna have to figure out what ports it's blocking, hopefully it's just the default RDP port.

Looking glass however is not working:

**Comment by T-vK** [comment]
Regarding looking glass did you set up the IVSHMEM server and socket like I did in the start-vm script before starting the looking glass client? Does the looking glass client have all permissions required to run properly? Have you installed the driver for the shared memory device within the VM?

3D applications have been a major issue for many people including me. But I remeber there was at least one guy who got it to work: https://github.com/jscinoz/optimus-vfio-docs/issues/2#issuecomment-663762701

Somewhere in this issue it's also mentioned how you can add a fake battery to the VM which fixed some nvidia driver issues within the VM. I can't remember if I ever added that to this project.

What notebook at you using btw? I should add it to the list of compatible notebooks on https://gpu-passthrough.com

I always assumed that iommu groups are kind of hardwired. It's interesting to hear that they are dynamic in a way. But I guess it makes sense, thinking about it there was a thing called ACS override patch that I always wanted to look into. From what I recall it allows you to split up iommu groups.

**Comment by mrkvn** [comment]
@Rabcor Do you have a working script or guide that I can follow to work with Arch/Manjaro? I have Dell Inspiron 7567, Nvidia 1050 GPU. I just follow guides. I don't do much debugging. So far, nothing has worked for me. I saw you in [this](https://gist.github.com/Misairu-G/616f7b2756c488148b7309addc940b28) gist too. Looks like you've been doing this for a long time now. For now, I just wanted the bare minimum instruction for it to work in order for me to know whether my laptop is able to actually do it. From the gist I linked, do you know what distro and what version the OP used? I just want to try the instructions as is, 'cause I'm not that good in changing the instruction that relates to the distro I'm using. For example, I tried to follow the instruction using Pop OS, which is not really different from the instruction 'cause he's using apt as package manager in the guide. But for example, he installed bumblebee, which I'm not sure is needed in Pop OS 'cause it already has GPU switching and I think Ubuntu has too, that's why I'm confused why bumblebee is needed. And also he's using Qemu v2 but I think it's v5 now so I'm not sure if I should use 2 or 5. Sorry for posting it here. I really need help :) . Of course, first thing I need to know is whether my laptop is supported. I'm pretty much sure it is but I can't know for sure if I'm not able to make it to work, which then might be the result of it "not being supported", so yeah, I'm in this loop of "unsureness". I consider myself a noob with these things, but I want to make it to work, at least until I confirm that it cannot work on my laptop.

**Comment by Rabcor** [comment]
Note: This comment is a work in progress and I will be editing it.

> @Rabcor Do you have a working script or guide that I can follow to work with Arch/Manjaro? I have Dell Inspiron 7567, Nvidia 1050 GPU. I just follow guides. I don't do much debugging. So far, nothing has worked for me. I saw you in [this](https://gist.github.com/Misairu-G/616f7b2756c488148b7309addc940b28) gist too. Looks like you've been doing this for a long time now. For now, I just wanted the bare minimum instruction for it to work in order for me to know whether my laptop is able to actually do it. From the gist I linked, do you know what distro and what version the OP used? I just want to try the instructions as is, 'cause I'm not that good in changing the instruction that relates to the distro I'm using. For example, I tried to follow the instruction using Pop OS, which is not really different from the instruction 'cause he's using apt as package manager in the guide. But for example, he installed bumblebee, which I'm not sure is needed in Pop OS 'cause it already has GPU switching and I think Ubuntu has too, that's why I'm confused why bumblebee is needed. And also he's using Qemu v2 but I think it's v5 now so I'm not sure if I should use 2 or 5. Sorry for posting it here. I really need help :) . Of course, first thing I need to know is whether my laptop is supported. I'm pretty much sure it is but I can't know for sure if I'm not able to make it to work, which then might be the result of it "not being supported", so yeah, I'm in this loop of "unsureness". I consider myself a noob with these things, but I want to make it to work, at least until I confirm that it cannot work on my laptop.

Sorry for the late response, I'm not very active on here, but sadly the answer is no, and if you "don't do much debugging" then this mostly experimental tech probably isn't gonna be very much for you, until you start 'doing much debugging' i spent weeks on and off over the span of months I mean look, I started trying to get this to work in june 2019 and the closest I ever got was "error code 43" in december 2020 and i felt like that was a huge achievement :D (even if it basically just proved that it probably couldn't be done with my laptop, but I hear nvidia has done something to make their windows drivers on recent windows 10 versions more compatible with this so it might be worth trying again, but that's not what I'm here for today)

Just saying, keep your expectations low, like real damn low. And be patient if you wanna try this stuff. I probably wouldn't ever have figured it out without T-vK's script :+1: 

Anyhow I am doing this again now on a fresh install of manjaro, I decided to properly ditch windows on my main machine and just set up a VM for running things I need but can't run on LInux (like possibly photoshop); for this I only intend to use the iGPU passthrough component of T-vK's Mobile Passthrough and skip over the entire dGPU passthrough part.

Anyhow, here is my adjusted dependency install line from manjaro (which will also work on arch)

`sudo pacman -S wget curl vim screen git lshw msr-tools sysfsutils remmina samba docker spice-gtk acpica cdrkit qemu virt-viewer`

Things u can only get from AUR (I use yay to download from AUR):

`yay -S virtio-win ovmf-git crudini uml_utilities uuid looking-glass`
Note: there is a package called edk2-ovmf however I am not sure if it is a suitable substitute for the ovmf package made specifically for qemu. Also, these aur packages might get moved to the official repos, be renamed or become outdated at any moment so install at your own risk.

intel specific:
`pacman -S intel-gpu-tools`
Note: could not find command `intel-virtual-output`, maybe because kernel params not yet set?


I did not find any suitable packages to replace qemu-utils, qemu-efi or qemu-kvm, only a qemu package, either all of these are included (seems somewhat likely to me) or I'm gonna need to do a lot of digging.

I do not intend to use bumblebee this time around, but if you're looking for ways to do that [this is how I succeeded in setting it up last time](https://forum.manjaro.org/t/how-to-install-up-to-date-bumblebee/41216), I'm sure this method is at least slightly outdated now though, but it might give you some idea of how to do it at least. I wait with bated breath for T-vK to figure out how to make this script compatible with Prime tho :laughing: 



For Vbios Finder:

pacman -S ruby ruby-bundler rubygems p7zip innoextract upx

Could not find: ruby-dev

Should try to run Vbios Finder script now without the apt-get command in it and see if it works, if not, I need to find a substitute for the ruby-dev package





For Kernel Params:

I do not feel like it is a very good idea to allow a script (especially one pulled from git and meant for another distro) to edit kernel params, solely because IF it fails for whatever reason, then it might break your system and you wouldn't be able to fix it again if you didn't know what the script was doing in the first place.

So I think manually editing kernel params is always gonna be the way to go.

These are the recommended kernel params from the script:

`iommu=1 amd_iommu=on intel_iommu=on i915.enable_gvt=1 kvm.ignore_msrs=1 rd.driver.pre=vfio-pci`
Note: 1 and on are not documented options for iommu and amd_iommu, I am as confused as T-vK about this...

Other kernel params not added in the script but recommended from [this guide](https://gist.github.com/Misairu-G/616f7b2756c488148b7309addc940b28)

`nogpumanager acpi_osi=! acpi_osi=Linux acpi_osi=\"Windows 2015\" pcie_port_pm=off`
(acpi_osi to windows 2009 if 2015 is disabling your trackpad)

also:

`default_hugepagesz=1G hugepagesz=1G hugepages=8 transparent_hugepage=never`

Warning: I had some issues with these parameters before, namely that 8x1G size non-transparent hugepages will eat up 8gb of RAM just by existing, a workaround for this is to disable transparent hugepages but disable hugepages (hugepages=0 instead of 8) and then enable hugepages via command before launching your VM like so:
`sudo sysctl -w vm.nr_hugepages=8`

and to disable again after turning off the vm:
`sudo sysctl -w vm.nr_hugepages=0`

Enabling non-transparent hugepages should help performance in your VM considerably, but it is essentially permanently allocating a set amount of RAM to VM use. (not sure if this info is outdated or not)

Every user should mix and match these kernel params as they seem fit it's part of the process, for instance I'm pretty sure I don't need amd_iommut on my intel laptop, but the acpi_osi params are probably likely to have some effect for me (possibly in ways entirely unrelated to the passthrough thing too)

I am not sure if I need nogpumanager and pcie_port_pm=off, not even sure what they do, but I will be using the scheme I suggested with hugepages (in the sense that I will use 1gb size hugepages and turn them off and on on-demand whenever I turn on my VM.

Now all the scripts in the utils directory except ones reliant on bumblebee should be working and I am currently testing that.

**Comment by T-vK** [comment]
I'm currently working torwards making this project distribution-agnostic. You might wanna check out the unattended-win-install branch. 
Dependecies aren't listed as package names anymore in that branch (see requirements.sh), but instead as binary names and files, which are then used to automatically scan the available package manager for the right packages to install.
But as I said, it's a work-in-progress. 
So far I have only added support for dnf and apt (although apt support hasn't even been tested yet). But adding more package managers should be relatively straight forward once I'm done.

**Comment by Rabcor** [comment]
@T-vK that is cool, can't wait to see it.

Also I updated my other post now, I reached the end and tried running a vm but it failed and I don't know why, I'd appreciate your input if you have any, I tested building qemu from git but I got the same result.

**Comment by T-vK** [comment]
Have you tried disabling dmabuf and just using qxl?

**Comment by Rabcor** [comment]
> Have you tried disabling dmabuf and just using qxl?

I did, and it runs. I can now get in with spice, but incidentally I just figured I might not necessarily need spice but I need RDP and RDP is not working (e.g. the xfreerdp command mentioned earlier in this thread); at least if I recall correctly RDP is the only way to get decent performance without using looking glass, which I can't use cuz I don't have an external display and unlike last time around it's not a firewall issue

Btw I found a [related wiki entry](https://wiki.archlinux.org/title/Intel_GVT-g#Using_DMA-BUF_with_UEFI/OVMF) for it now that I knew what to look for, so thanks :)

**Comment by T-vK** [comment]
@Rabcor Out of curiosity, you said that `service docker start needs to be changed to systemctl start docker`. Is that true? I'm just asking because many distributions allow both (if the system uses systemd, it will internally translate that to a systemctl call). It seemed like the most portable way of starting a service, but if arch doesn't support it, I need to add my own abstraction layer.
